packages/api/src/promise/index.ts

```diff
@@ -195,17 +197,42 @@ export default class ApiPromise extends ApiBase<DecoratedRpc, QueryableStorage,
         rxfn(...params).toPromise();
     }
 
-    return (..._params: Array<any>): UnsubFunction => {
+    return (..._params: Array<any>): Promise<UnsubFunction> => {
       const cb = _params[_params.length - 1];
 
       assert(isFunction(cb), 'Expected callback as last paramater for subscription');
 
-      const params = _params.slice(0, _params.length - 1);
-      const subscription = rxfn(...params).subscribe(cb);
-
-      return (): void => {
-        subscription.unsubscribe();
-      };
+      return new Promise((resolve, reject) => {
+        let isCompleted = false;
+        const params = _params.slice(0, _params.length - 1);
+        const subscription = rxfn(...params)
+          .pipe(
+            // if we find an error (invalid params, etc), reject the promise
+            catchError((error) => {
+              console.error(error);
+
+              if (!isCompleted) {
+                isCompleted = true;
+
+                reject(error);
+              }
+
+              // we don't want to continue, so empty observable it is
+              return EMPTY;
+            }),
+            // upon the first result, resolve the with the unsub function
+            map((result: any) => {
```

Comment:
Note sure about this part at all - but want to resolve only once we are certain there are no errors...